source/drivers/HIDKeyboard.cpp

```diff
@@ -0,0 +1,409 @@
+    0x95, 0x10,                         // Report Count (16)
+    0x09, 0xe2,                         // Usage (Mute) 0x01
+    0x09, 0xe9,                         // Usage (Volume Up) 0x02
+    0x09, 0xea,                         // Usage (Volume Down) 0x03
+    0x09, 0xcd,                         // Usage (Play/Pause) 0x04
+    0x09, 0xb7,                         // Usage (Stop) 0x05
+    0x09, 0xb6,                         // Usage (Scan Previous Track) 0x06
+    0x09, 0xb5,                         // Usage (Scan Next Track) 0x07
+    0x0a, 0x8a, 0x01,                   // Usage (Mail) 0x08
+    0x0a, 0x92, 0x01,                   // Usage (Calculator) 0x09
+    0x0a, 0x21, 0x02,                   // Usage (www search) 0x0a
+    0x0a, 0x23, 0x02,                   // Usage (www home) 0x0b
+    0x0a, 0x2a, 0x02,                   // Usage (www favorites) 0x0c
+    0x0a, 0x27, 0x02,                   // Usage (www refresh) 0x0d
+    0x0a, 0x26, 0x02,                   // Usage (www stop) 0x0e
+    0x0a, 0x25, 0x02,                   // Usage (www forward) 0x0f
+    0x0a, 0x24, 0x02,                   // Usage (www back) 0x10
+    0x81, 0x62,                         // Input (Data,Var,Abs,NPrf,Null)
+    0xc0,                               // End Collection
+
+};
+
+static const HIDReportDescriptor reportDesc = {
+    9,
+    0x21,                  // HID
+    0x101,                 // hidbcd 1.01
+    0x00,                  // country code
+    0x01,                  // num desc
+    0x22,                  // report desc type
+    sizeof(hidKeyboardDescriptor),
+};
+
+static const InterfaceInfo ifaceInfo = {
+    &reportDesc,
+    sizeof(reportDesc),
+    1,
+    {
+        1,    // numEndpoints
+        0x03, /// class code - HID
+        0x01, // subclass (boot interface)
+        0x01, // protocol (keyboard)
+        0x00, //
+        0x01, //
+    },
+    {USB_EP_TYPE_INTERRUPT, 1},
+    {USB_EP_TYPE_INTERRUPT, 1},
+};
+
+USBHIDKeyboard::USBHIDKeyboard() : USBHID()
+{
+    for(int i=0; i<HID_KEYBOARD_KEYSTATE_SIZE_GENERIC; i++)
+        keyStateGeneric[i] = HID_KEYBOARD_KEY_OFF;
+    
+    for(int i=0; i<HID_KEYBOARD_KEYSTATE_SIZE_CONSUMER; i++)
+        keyStateConsumer[i] = HID_KEYBOARD_KEY_OFF;
+
+    keyPressedCountGeneric = 0;
+    keyPressedCountConsumer = 0;
+}
+
+int USBHIDKeyboard::stdRequest(UsbEndpointIn &ctrl, USBSetup &setup)
+{
+    if (setup.bRequest == GET_DESCRIPTOR)
+    {
+        if (setup.wValueH == 0x21)
+        {
+            InterfaceDescriptor tmp;
+            fillInterfaceInfo(&tmp);
+            return ctrl.write(&tmp, sizeof(tmp));
+        }
+        else if (setup.wValueH == 0x22)
+        {
+            return ctrl.write(hidKeyboardDescriptor, sizeof(hidKeyboardDescriptor));
+        }
+    }
+    return DEVICE_NOT_SUPPORTED;
+}
+
+const InterfaceInfo *USBHIDKeyboard::getInterfaceInfo()
+{
+    return &ifaceInfo;
+}
+
+int USBHIDKeyboard::keyDown(uint8_t key, uint8_t reportID)
```

Comment:
Is `key` a character that is used as a look up? For instance, if I wanted to use this API, would I call it like this?

```
USBHIDKeyboard keyboard;

keyboard.keyDown('a');
keyboard.keyUp('a');
```

Would the result be that the character 'a' would appear on my screen if my cursor was focused on a text field?

Is it likely that a user would want to change the report ID?

Reply:
Perhaps there should be a `press(uint8_t key)` API as well!

Reply:
The uint8_t key does not match up with char 'a'.  These codes will be different depending on the keyboard layout selected on the host machine. In emailing with @mmoskal, he had told us

>IMHO the more we put in TypeScript the better especially for more esoteric features. The C++ code generally gets linked every time whereas Typescript is only linked as needed>

For this reason the current C++ API will just feed through whatever code is passed, and leave the keycode lookup to typescript.

here is a reference for keycodes: https://gist.github.com/MightyPork/6da26e382a7ad91b5496ee55fdc73db2

if you did
`USBHIDKeyboard keyboard;

keyboard.keyDown(0x04);
keyboard.keyUp(0x04);`

then the character 'a' would appear in whatever text field was focused.

The alternate report ID is to support multimedia keys like play, stop, skip song, etc.

I can add a `press(uint8_t key)` wrapper as well with builtin delay.

Reply:
Ok, @mmoskal can you elaborate? To me it would make sense to just put the lookup in C/C++, and store the desired characters in flash like in the Bluetooth HID class for the micro:bit:

https://github.com/lancaster-university/microbit-dal/commit/63bbb2c2982724fb9a5640364f6f5f945535814f#diff-fe9563eac04d33a4e76276c520206ecfR186

Reply:
As far as I understand, you need different translation tables for different languages. I would rather have them in flash from TypeScript, where I can pick one, not all.

If you do build in translation tables, please make sure they are not linked when not referenced.

Reply:
it mostly depends on localization plans. if y'all never plan to do anything but US Keyboard layout, this stuff becomes very easy. but the moment you want it to work with anything else, you need a translation table. traditionally, these tables are handled at a higher level, so you can more easily swap them around, then you just index as needed.

see for example, javascript: https://www.w3schools.com/jsref/event_key_keycode.asp
or Golang: https://godoc.org/golang.org/x/mobile/event/key

US ascii-to-keycode in circuitpython: https://github.com/adafruit/Adafruit_CircuitPython_HID/blob/master/adafruit_hid/keyboard_layout_us.py

Reply:
Actually, in MakeCode we could ask the user to press the key and tell them the code (reading the value from keyboard event). This would be particularly useful for various multimedia keys etc.

In case of MakeCode, the translation tables for different keyboard layouts should come from people who actually speak the language, probably as external packages. Another issue is that MakeCode doesn't currently generate Unicode strings. 

We could have a US keymap in Codal, but then why not UK, or German, or whatever. If interested, you can take a look at the ~200 layouts here: http://http.debian.net/debian/pool/main/c/console-data/console-data_1.12.orig.tar.bz2 (and I think these are the actually used ones).

Reply:
@mmoskal That sounds like a good idea for multimedia keys, what's the plan for alphanumeric characters? Transpose them into keyboard values at compile time?

@ladyada I'd love to see at least one keymap in Codal (which I'm more than happy to contribute). It makes me sad that I can't write:
```
cplay.keyboard.type("Hello");
```
What if we were to provide a weak symbol for the keymap, with a sensible default, which could be overridden by an external package, for example?

Would that be workable?

Reply:
sure, we can do US keyboard layout, it would be a layer on top of this API. how about we merge this and then have another PR for the next-up layer?

Reply:
@ladyada Agreed! :smile:

Reply:
yay - thank you :)

for the overlying layer, whats a desired API:
           cplay.keyboard.type(char *, keymap *)
           cplay.keyboard.keyPress(char c, keymap *), keyRelease and modifierKeyPress/Release, mediaKeyPress/Release + releaseAll()

or maybe support unicode if theres an ascii->unicode converter? anything else?

Reply:
@ladyada I think something a bit like that would be great. After chatting with @mmoskal, to be able to cleanly allow other key mappings, we realised that a key map abstraction would be required. This abstraction would have to be able to map ascii and unicode characters (as you identified) to a sequence of key presses, for example in mandarin, some characters translate into multiple key presses. 

Further, users should be able to add their own keyboards which requires a common interface for mapping keys (the KeyMap class):
```cpp
struct Key 
{
    uint16_t code;
    uint16_t modifier; 
}

class KeyMap
{
    KeyMap();

    // return a sequence of Keys in a ManagedBuffer
    ManagedBuffer mapCharacter(uint16_t c)
    {
        // some mapping defined by the user who adds the keyboard.
    }
}
```
In addition, users should be able to override the default KeyMap from MakeCode using a C++ API. For the USBHIDKeyboard, I think something that looks a bit like this:
```cpp
static KeyMap* currentMap = UKKeyMap; 

// add our sink node class for streams.
class USBHIDKeyboard : public SinkNode
{
    USBHIDKeyboard(KeyMap& m);

    // sets currentMap
    int setKeyMap(KeyMap& m);

    // keep these because they are useful for advanced users, and as internal functions
    int modifierKeyDown(uint16_t key, uint8_t reportID=HID_KEYBOARD_REPORT_GENERIC);
    int modifierKeyUp(uint16_t key, uint8_t reportID=HID_KEYBOARD_REPORT_GENERIC);
    int keyDown(uint16_t key, uint8_t reportID=HID_KEYBOARD_REPORT_GENERIC);
    int keyUp(uint16_t key, uint8_t reportID=HID_KEYBOARD_REPORT_GENERIC);

    // automatic inference of required key presses using currentMap, followed by KeyDown etc.
    int press(uint16_t c);
    // type is a sequence of presses
    int type(uint16_t* c);
    int type(ManagedString s);

    // releases any keys pressed.
    int release();
}
```
Another thing to consider is multiple fibers (the lightweight threads we run in CODAL) accessing the HIDKeyboard interface in quick succession. I assume @deanm1278, there is the possibility of the key press buffer filling up?  In which case, the queueing mechanisms offered by our stream interface would be perfect here.

We also need to support unicode in `ManagedString` which should definitely come at the same time as these changes.

Thoughts? Anything I missed?